source/common/upstream/eds.cc

```diff
@@ -1,53 +1,54 @@
-#include "common/upstream/sds.h"
+#include "common/upstream/eds.h"
 
-#include <cstdint>
-#include <string>
-#include <vector>
-
-#include "common/http/headers.h"
-#include "common/json/config_schemas.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
+#include "common/upstream/sds_subscription.h"
 
 namespace Envoy {
 namespace Upstream {
 
-SdsClusterImpl::SdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
+EdsClusterImpl::EdsClusterImpl(const Json::Object& config, Runtime::Loader& runtime,
                                Stats::Store& stats, Ssl::ContextManager& ssl_context_manager,
                                const SdsConfig& sds_config, const LocalInfo::LocalInfo& local_info,
                                ClusterManager& cm, Event::Dispatcher& dispatcher,
                                Runtime::RandomGenerator& random)
-    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager),
-      RestApiFetcher(cm, sds_config.sds_cluster_name_, dispatcher, random,
-                     sds_config.refresh_delay_),
-      local_info_(local_info), service_name_(config.getString("service_name")) {}
-
-void SdsClusterImpl::parseResponse(const Http::Message& response) {
-  Json::ObjectSharedPtr json = Json::Factory::loadFromString(response.bodyAsString());
-  json->validateSchema(Json::Schema::SDS_SCHEMA);
+    : BaseDynamicClusterImpl(config, runtime, stats, ssl_context_manager), local_info_(local_info),
+      cluster_name_(config.getString("service_name")) {
+  // TODO(htuch): This is where the v2 Subscription implementations will be constructed based on
+  // config. Today, we just reuse the v1 SDS config and an adapter Subscription to the v1 REST JSON
+  // API.
+  subscription_.reset(new SdsSubscription(info_->stats(), sds_config, cm, dispatcher, random));
+}
+
+void EdsClusterImpl::initialize() { subscription_->start({cluster_name_}, *this); }
+
+bool EdsClusterImpl::onConfigUpdate(const ResourceVector& resources) {
```

Comment:
I think we can do stats/logging consistently in the Subscription implementation (whether it be SDS, {Grpc,Filesystem,Http}SubscriptionImpl) regardless of exception vs. bool return, and that's the eventual plan. I think the one thing that exception catching buys you is that if we call into common code in `onConfigUpdate()` that is shared with startup time configuration and that expects to be able to throw exceptions, we don't need to deal with both a false return and exception catch in each of the Subscription implementations.

Being a Googler, I'm naturally allergic to exceptions, but I think it fits the Envoy model today if you agree with the above. WDYT?

I'm not fully understanding your second comment, can you elaborate? Thanks.